src/EditorFeatures/CSharpTest/Completion/CompletionProviders/PropertySubPatternCompletionProviderTests.cs

```diff
@@ -0,0 +1,552 @@
+        {
+            var markup =
+@"
+class Program
+{
+    void M()
+    {
+        _ = this is Derived { $$ }
+    }
+}
+class Derived : Program
+{
+    private int P1 { get; set; }
+    private int P2 { get; set; }
+}
+";
+            await VerifyNoItemsExistAsync(markup);
+        }
+
+        [Fact]
+        public async Task PropertiesInRecursivePattern_WithDerivedType_WithInaccessibleMembers2()
+        {
+            var markup =
+@"
+class Program
+{
+    private int P1 { get; set; }
+    void M()
+    {
+        _ = this is Derived { $$ }
+    }
+}
+class Derived : Program
+{
+    private int P2 { get; set; }
+}
+";
+            await VerifyItemExistsAsync(markup, "P1");
+        }
+
+        [Fact]
+        public async Task PropertiesInRecursivePattern_UseStaticTypeFromIs()
+        {
+            var markup =
+@"
+class Program
+{
+    public int P1 { get; set; }
+    public int P2 { get; set; }
+
+    void M()
+    {
+        _ = this is { $$ }
+    }
+}
+";
+            // VerifyItemExistsAsync also tests with the item typed.
+            await VerifyItemExistsAsync(markup, "P1");
+            await VerifyItemExistsAsync(markup, "P2");
+        }
+
+        [Fact]
+        public async Task PropertiesInRecursivePattern_InSwitchStatement()
+        {
+            var markup =
+@"
+class Program
+{
+    public int P1 { get; set; }
+    public int P2 { get; set; }
+
+    void M()
+    {
+        switch (this)
+        {
+            case Program { $$ }
+        }
+    }
+}
+";
+            // VerifyItemExistsAsync also tests with the item typed.
+            await VerifyItemExistsAsync(markup, "P1");
+            await VerifyItemExistsAsync(markup, "P2");
+        }
+
+        [Fact]
+        public async Task PropertiesInRecursivePattern_UseStaticTypeFromSwitchStatement()
+        {
+            var markup =
+@"
+class Program
+{
+    public int P1 { get; set; }
+    public int P2 { get; set; }
+
+    void M()
+    {
+        switch (this)
+        {
+            case { $$ }
+        }
+    }
+}
+";
+            // VerifyItemExistsAsync also tests with the item typed.
+            await VerifyItemExistsAsync(markup, "P1");
+            await VerifyItemExistsAsync(markup, "P2");
+        }
+
+        [Fact(Skip = "PROTOTYPE(patterns2)")]
+        public async Task PropertiesInRecursivePattern_InSwitchExpression()
+        {
+            // PROTOTYPE(patterns2): I think this will be fixed by using GetTypeInfo (once merged)
+            var markup =
+@"
+class Program
+{
+    public int P1 { get; set; }
+    public int P2 { get; set; }
+
+    void M()
+    {
+        _ = this switch { { $$ } }
+    }
+}
+";
+            await VerifyItemExistsAsync(markup, "P1");
+            await VerifyItemExistsAsync(markup, "P2");
+        }
+
+        [Fact(Skip = "PROTOTYPE(patterns2)")]
+        public async Task PropertiesInRecursivePattern_WithPositionalPattern()
```

Comment:
Please fix these cases. You need to take into account that the parent of a PropertySubpattern can also be a DeconstructionPattern, not just PropertyPattern, everywhere you expect that to be the case.

See https://sharplab.io/#v2:EYLgZgpghgLgrgJwgZwLRIMaOQSwG4SoAOsMECAdsgDQwhTIzUAmIA1AD4ACADAARcAjAG4AsACguAZgEAmPgGE+Abwl91AmVwAsfALIAKAJRqNq8Rst8A+nwC8fGAAscyPq74HB1PrKMq+AAVBED5vINlQ+QBfMQsNaNN1JPcKGCDBAIBzCBhhPkT49Rw0iOzc/MLLFJ0+ABEIDAB7KhgEOAwYAya4dJL0qB8evtLgEyKVFMsoez5gWZ44y0LooA===

then make sure it works both with and without a type:
```c#
this is C (1, 2) { P1: 1, P2: 2 };
this is (1, 2) { P1: 1, P2: 2 };
```

Reply:
I don't plan to expand the cases handled by the completion provider in this PR. As noted, we have a design to supercede this approach, using [GetTypeInfo](https://github.com/dotnet/roslyn/pull/26521). I'm waiting for that to be merged.

Reply:
Note: even when you do that, you will still have to update the code to get the DeconstructionPattern in order to use GetTypeInfo on it. Right now I found 2 places in the code that assume the parent can only be a PropertyPattern and bail out if it's not.